src/io/download.jl

```diff
@@ -1,78 +1,71 @@
-            error("The file \"$filename\" was not found in \"$dir\". ",
-                  "You can download the dataset at $(settings.url), ",
-                  "or try again from the REPL in order for ",
-                  "`MLDatasets` to download it for you.")
+            # use user-provided dir
+            if isdir(dir)
+                dir
+            else
+                !DataDeps.env_bool("DATADEPS_DISABLE_DOWNLOAD") || error("DATADEPS_DISABLE_DOWNLOAD enviroment variable set. Can not trigger download.")
+                DataDeps.download(DataDeps.registry[depname], dir)
+                dir
+            end
         end
-    end
-    path
+    end::String
 end
 
-"""
-    download_helper(dir, baseurl, files; [i_accept_the_terms_of_use = false], [prompt])
-
-Check if the `files` are present in the specified `dir`, or if
-any of the files are missing. In the case that any of the `files`
-are missing and `i_accept_the_terms_of_use=false` the function
-will raise a warning or an error depending on if julia is run in
-an interactive session. If an interactive session is detected the
-user will be presented with `prompt` and the option to
-download the dataset to the specified `dir`.
-"""
-function download_helper(
-        dir::AbstractString,
-        baseurl::AbstractString,
-        files::AbstractVector;
-        i_accept_the_terms_of_use = false,
-        prompt = error("Please provide a \"prompt\" for the given dataset"))
-    missing = filter(file->!isfile(joinpath(dir, file)), files)
-    if !isempty(missing)
-        info("The specified directory \"$dir\" is missing the files ",
-             join(map(f->"\"$f\"", missing), ", ", " and "),
-             "of the full data set.")
-        if !i_accept_the_terms_of_use && isinteractive()
-            prompt = "\n" * prompt * "\nDo you want to download the dataset from $baseurl to \"$dir\"? [y/n] "
-            print(prompt)
-            answer = first(readline())
-            if answer == 'y'
-                i_accept_the_terms_of_use = true
-            end
-        end
-        if i_accept_the_terms_of_use
-            mkpath(dir)
-            for file in missing
-                url = baseurl * "$file"
-                path = joinpath(dir, file)
-                info("downloading $file from $url to $dir")
-                run(BinDeps.download_cmd(url, path))
-            end
+function datafile(depname, filename, dir = nothing; recurse = true, kw...)
+    path = joinpath(datadir(depname, dir; kw...), filename)
+    if !isfile(path)
+        warn("The file \"$path\" does not exist, even though the dataset-specific folder does. This is an unusual situation that may have been caused by a manual creation of an empty folder, or manual deletion of the given file \"$filename\".")
+        info("Retriggering DataDeps.jl for \"$depname\" to \"$dir\".")
+        download_dep(depname, dir; kw...)
+        if recurse
+            datafile(depname, filename, dir; recurse = false, kw...)
```

Comment:
It is unclear to me what `recurse` does.
I think maybe this line should be:
`datafile(depname, filename, path; recurse = true, kw...)` ?

Reply:
This is basically just to avoid an "endless loop" situation if you have a type in the filename. Because if you do, then the download will be triggered (because the file isn't found) and then this function is called again. since the file still can't be found the download would be triggered again, and so on